In [1]:
import pandas as pd
pd.options.display.float_format = '{:.8e}'.format
from population import generate_init
from character import CharacterType
#from src.stats import random_stats

N = 200

maxStatsValue = 150
type = CharacterType.ARCHER

#equipment_data = random_stats(N, maxStatsValue)
poblacion_inicial = generate_init(N,maxStatsValue,type)

#poblacion_inicial


In [2]:
#Añado las columnas de desempeño relativo y acumulado
from selection import add_relative_accumulate
poblacion_rel_acu = add_relative_accumulate(poblacion_inicial)
poblacion_rel_acu

sum: 54267.131422148566


strength  agility  expertise  resistance  life         height  \
31         57        3         74          14     2 2.00000000e+00   
140        90        3         49           2     6 1.50000000e+00   
20         60        1         61          24     4 1.70000000e+00   
64         63       29         48           3     7 1.80000000e+00   
161        66        9         50           8    17 2.00000000e+00   
..        ...      ...        ...         ...   ...            ...   
199         6       18          8          81    37 1.50000000e+00   
183         4       36          6          43    61 1.30000000e+00   
137        21        5          2         110    12 1.40000000e+00   
29          2       19         21         100     8 1.50000000e+00   
51          6        2          2         100    40 2.00000000e+00   

                     type    performance  performance_relative  \
31   CharacterType.ARCHER 1.05965298e+03        1.95266077e-02   
140  CharacterType.ARCHER 1.05639484e+03        1.94665687e-02   
20   CharacterType.ARCHER 9.29123050e+02        1.71212855e-02   
64   CharacterType.ARCHER 8.98809480e+02        1.65626864e-02   
161  CharacterType.ARCHER 8.98796105e+02        1.65624399e-02   
..                    ...            ...                   ...   
199  CharacterType.ARCHER 2.53616792e+01        4.67348808e-04   
183  CharacterType.ARCHER 2.03594968e+01        3.75171789e-04   
137  CharacterType.ARCHER 1.99485847e+01        3.67599765e-04   
29   CharacterType.ARCHER 1.69754882e+01        3.12813443e-04   
51   CharacterType.ARCHER 7.87950771e+00        1.45198530e-04   

     performance_accumulated  
31            1.95266077e-02  
140           3.89931764e-02  
20            5.61144619e-02  
64            7.26771483e-02  
161           8.92395882e-02  
..                       ...  
199           9.98799216e-01  
183           9.99174388e-01  
137           9.99541988e-01  
29            9.99854801e-01  
51            1.00000000e+00  

[200 rows x 10 columns]

#### Seleccion de k indiduos
A*(método1) + (1-A)*(método2)

In [3]:
import numpy as np
#Dado un K
#Si quiero un m1_perc de la selección con metodo1
K = 120
m1_perc = 0.3
N1 = int(np.round(K * m1_perc))
N2 = K - N1
#N1

In [4]:
from selection import elitist_selection
nueva_poblacion_e = elitist_selection(poblacion_rel_acu,N1)
#nueva_poblacion_e

In [5]:
from selection import roulette_wheel_selection
nueva_poblacion_rul = roulette_wheel_selection(poblacion_rel_acu, N2)
#nueva_poblacion_rul

In [6]:
import pandas as pd
selection_total = pd.concat([nueva_poblacion_e,nueva_poblacion_rul])
selection_total

strength agility expertise resistance life         height  \
0        57       3        74         14    2 2.00000000e+00   
1        90       3        49          2    6 1.50000000e+00   
2        60       1        61         24    4 1.70000000e+00   
3        63      29        48          3    7 1.80000000e+00   
4        66       9        50          8   17 2.00000000e+00   
..      ...     ...       ...        ...  ...            ...   
79       22       4        74         10   40 1.60000000e+00   
80       32       6        63         15   34 1.80000000e+00   
81       11      29        59         22   29 1.80000000e+00   
82       26      34        24         24   42 1.50000000e+00   
83       70      51        15         10    4 1.60000000e+00   

                    type    performance performance_relative  \
0   CharacterType.ARCHER 1.05965298e+03       1.95266077e-02   
1   CharacterType.ARCHER 1.05639484e+03       1.94665687e-02   
2   CharacterType.ARCHER 9.29123050e+02       1.71212855e-02   
3   CharacterType.ARCHER 8.98809480e+02       1.65626864e-02   
4   CharacterType.ARCHER 8.98796105e+02       1.65624399e-02   
..                   ...            ...                  ...   
79  CharacterType.ARCHER 4.51118398e+02       8.31292140e-03   
80  CharacterType.ARCHER 5.83956580e+02       1.07607785e-02   
81  CharacterType.ARCHER 2.15213815e+02       3.96582257e-03   
82  CharacterType.ARCHER 2.43460807e+02       4.48634009e-03   
83  CharacterType.ARCHER 4.33388342e+02       7.98620327e-03   

   performance_accumulated  
0           1.95266077e-02  
1           3.89931764e-02  
2           5.61144619e-02  
3           7.26771483e-02  
4           8.92395882e-02  
..                     ...  
79          4.55618558e-01  
80          2.93601671e-01  
81          7.88832436e-01  
82          7.30378383e-01  
83          4.63604761e-01  

[120 rows x 10 columns]

In [7]:
from genes import  encode_genes

cromosomas = encode_genes(selection_total)
#cromosomas

In [8]:
from crossover import single_point_crossover, valid_chromosome

new_individuals = single_point_crossover(cromosomas)
cromosomas_ok1 = valid_chromosome(new_individuals)
while (len(cromosomas_ok1)<1):
    new_individuals = single_point_crossover(cromosomas)
    cromosomas_ok1 = valid_chromosome(new_individuals)
#new_individuals
cromosomas_ok1

strength   agility expertise resistance      life height
0    00111010  00000011  01001001   00001010  00000110    110
1    01011001  00000011  00110010   00000110  00000010    011
12   00111101  00001000  00100100   00100001  00001100    110
13   00111000  00001110  00101000   00011010  00001110    111
40   00100000  00000110  00111111   00001111  00100010    101
41   00100000  00000110  00111111   00001111  00100010    101
60   00110100  00000110  01001111   00001011  00000010    110
61   00111001  00000011  00111010   00001110  00010010    001
108  01000001  00100110  00101100   00000010  00000001    010
109  01000001  00000011  00010010   00011011  00100101    011

In [9]:
#from genes import decode_genes
#cromosomas_decoded = decode_genes(new_individuals)
#cromosomas_decoded

- Cantidad de hijos validos: 4
-     strength  agility expertise	resistance	life  height  type	                performance	 performance_relative   performance_accumulated
- 5	  2	        30	    3	        6	        109	  2.0	  CharacterType.ARCHER	6.277699	 0.001625	            1.0
- 6	  50	    18	    3	        18	        61	  1.7	  CharacterType.ARCHER	119.533975	 0.030933	            0.881345
- 9	  4	        39	    92	        4	        11	  1.4	  CharacterType.ARCHER	106.010218	 0.027433	            0.908778
- 10  4	        39	    92	        4	        11	  1.4	  CharacterType.ARCHER	106.010218	 0.027433	            0.908778

In [10]:
from crossover import two_point_crossover, valid_chromosome

new_individuals2 = two_point_crossover(cromosomas)
cromosomas_ok2 = valid_chromosome(new_individuals2)

while (len(cromosomas_ok2)<1):
    new_individuals2 = single_point_crossover(cromosomas)
    cromosomas_ok2 = valid_chromosome(new_individuals2)
#new_individuals2
cromosomas_ok2

strength   agility expertise resistance      life height
40   00100000  00000110  00111111   00001111  00100010    101
41   00100000  00000110  00111111   00001111  00100010    101
100  00111111  00011101  00110000   00000011  00000111    101
101  00101111  00100000  00001011   00011010  00100010    001
104  00001001  00011101  00111000   00100111  00010001    011
105  01000001  00000011  00010010   00011010  00100110    000

In [11]:
#from genes import decode_genes
#cromosomas_decoded2 = decode_genes(new_individuals2)
#cromosomas_decoded2

In [12]:
#from genes import decode_genes
#cromosomas_decoded2 = decode_genes(cromosomas_ok)
#cromosomas_decoded2

In [13]:
from mutation import one_gene_mutation
from crossover import valid_chromosome

mutation_rate = 0.1
if( len(cromosomas_ok1) > len(cromosomas_ok2) ):

    print("single_point_crossover won")
    result = one_gene_mutation(cromosomas_ok1, mutation_rate)
    cromosomas_okm = valid_chromosome(result)

    while (len(cromosomas_okm)<1):
        result = one_gene_mutation(cromosomas_ok1, mutation_rate)
        cromosomas_okm = valid_chromosome(result)
else:     
    print("two_point_crossover won")   
    result = one_gene_mutation(cromosomas_ok2, mutation_rate)
    cromosomas_okm = valid_chromosome(result)

    while (len(cromosomas_okm)<1):
        result = one_gene_mutation(cromosomas_ok2, mutation_rate)
        cromosomas_okm = valid_chromosome(result)


cromosomas_okm



single_point_crossover won


strength   agility expertise resistance      life height
40   00100000  00000110  00111111   00001111  00100010    101
61   00111001  00000011  00111010   00001110  00010010    001
108  01000001  00100110  00101100   00000010  00000001    010

In [14]:
from genes import decode_genes
cromosomas_decoded2 = decode_genes(cromosomas_okm)
cromosomas_decoded2

strength  agility  expertise  resistance  life         height
1        32        6         63          15    34 1.80000000e+00
2        57        3         58          14    18 1.40000000e+00
3        65       38         44           2     1 1.50000000e+00

In [15]:
from mutation import multi_gene_mutation

# Ejemplo de uso:
mutation_rate = 0.8
resultmp = multi_gene_mutation(cromosomas_ok1, mutation_rate)
resultmp



   strength   agility expertise resistance      life height
0  00111010  00000011  01001001   00001010  00000110    110
1  01011001  00000011  00110010   00000110  00000010    011
2  00111101  00001000  00100100   00100001  00001100    110
3  00111000  00001110  00101000   00011010  00001110    111
4  00100000  00000110  00111111   00001111  00100010    101
5  00100000  00000110  00111111   00001111  00100010    101
6  00110100  00000110  01001111   00001011  00000010    110
7  00111001  00000011  00111010   00001110  00010010    001
8  01000001  00100110  00101100   00000010  00000001    010
9  01000001  00000011  00010010   00011011  00100101    011


strength   agility expertise resistance      life height
0  11000110  01111100  10110010   11110110  11011001    001
1  10110100  00000011  00110010   01101000  11111101    011
2  01010010  01110111  11011111   00100001  00001100    001
3  10001111  11110010  00101000   11101100  11100001    100
4  11011111  11111011  00111111   00110000  11011001    111
5  11011111  00000110  11000010   11111010  00100010    010
6  11011011  01111001  10111010   11110100  11011101    000
7  11000010  10110110  01000110   01110001  10101101    001
8  10111101  11011010  10000001   11111011  00000001    100
9  01000001  01110100  11000101   10101000  11010010    100

In [33]:
from mutation import multi_gene_mutation_uniform

# Ejemplo de uso:
mutation_rate = 0.7
resultmu = multi_gene_mutation_uniform(cromosomas_ok1, mutation_rate)
resultmu


ImportError: cannot import name 'multi_gene_mutation_uniform' from 'mutation' (c:\Users\dpenaloza\Downloads\SIA\sia-tp2\src\mutation.py)

In [16]:
from population import eval_performace

new_generation = eval_performace(cromosomas_decoded2,type)
new_generation

strength  agility  expertise  resistance  life         height  \
2        57        3         58          14    18 1.40000000e+00   
3        65       38         44           2     1 1.50000000e+00   
1        32        6         63          15    34 1.80000000e+00   

                   type    performance  
2  CharacterType.ARCHER 8.77092033e+02  
3  CharacterType.ARCHER 8.69243777e+02  
1  CharacterType.ARCHER 5.83956580e+02

In [17]:
import numpy as np
from selection import boltzmann_selection

# Example usage:

max_generation = 18
generation = 1
k = 4                                  # Number of desired selected individuals.
T_0 = 20                                 # Initial temperature.
T_offset = 0.5                          # Offset value for temperature exponential function.
reference_generation = max_generation   # Reference generation where 0 < (T_0 - T_offset) = 0.015 = deltaT is desired.
deltaT = 0.015                          # 0 < (T_0 - T_offset) = deltaT at reference generation.
m = - (np.log(deltaT) - np.log(T_0 - T_offset)) / reference_generation

new_boltzmann_selection = boltzmann_selection(poblacion_inicial,k,generation,T_0,T_offset,m)

new_boltzmann_selection


sum: 1.0
     strength  agility  expertise  resistance  life         height  \
31         57        3         74          14     2 2.00000000e+00   
140        90        3         49           2     6 1.50000000e+00   
20         60        1         61          24     4 1.70000000e+00   
64         63       29         48           3     7 1.80000000e+00   
161        66        9         50           8    17 2.00000000e+00   
..        ...      ...        ...         ...   ...            ...   
199         6       18          8          81    37 1.50000000e+00   
183         4       36          6          43    61 1.30000000e+00   
137        21        5          2         110    12 1.40000000e+00   
29          2       19         21         100     8 1.50000000e+00   
51          6        2          2         100    40 2.00000000e+00   

                     type    performance  performance_relative  \
31   CharacterType.ARCHER 5.59612127e-01        5.59612127e-01   
140  CharacterType

strength  agility  expertise  resistance  life         height  \
0        57        3         74          14     2 2.00000000e+00   
1        57        3         74          14     2 2.00000000e+00   
2        90        3         49           2     6 1.50000000e+00   
3        90        3         49           2     6 1.50000000e+00   

                   type    performance  performance_relative  \
0  CharacterType.ARCHER 5.59612127e-01        5.59612127e-01   
1  CharacterType.ARCHER 5.59612127e-01        5.59612127e-01   
2  CharacterType.ARCHER 4.40341485e-01        4.40341485e-01   
3  CharacterType.ARCHER 4.40341485e-01        4.40341485e-01   

   performance_accumulated  
0           5.59612127e-01  
1           5.59612127e-01  
2           9.99953612e-01  
3           9.99953612e-01

In [18]:
from selection import deterministic_tournament_selection 

individualsT = deterministic_tournament_selection(poblacion_inicial, 4, 5)
individualsT

c:\Users\dpenaloza\Downloads\SIA\sia-tp2\src\selection.py:152: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.
  individuals = pd.concat([individuals, winner])


strength agility expertise resistance life         height  \
0       11      49        61          3   26 2.00000000e+00   
1        8      47        42         40   13 1.80000000e+00   
2       36      42        41         18   13 1.60000000e+00   
3       46      23        58          3   20 1.40000000e+00   
4       20      12        84         22   12 1.30000000e+00   

                   type    performance  
0  CharacterType.ARCHER 2.28399064e+02  
1  CharacterType.ARCHER 1.25699387e+02  
2  CharacterType.ARCHER 4.98570087e+02  
3  CharacterType.ARCHER 7.78440493e+02  
4  CharacterType.ARCHER 4.14160318e+02

In [19]:
from selection import stochastic_tournament_selection
#ejemplo
individualsTd = stochastic_tournament_selection(poblacion_inicial, 4)
individualsTd

strength  agility  expertise  resistance  life         height  \
181         8       51         59           1    31 1.60000000e+00   
163        27        8         25          40    50 1.30000000e+00   
150        12       76         15          45     2 1.60000000e+00   
89         22        4         10          81    33 1.30000000e+00   

                     type    performance  
181  CharacterType.ARCHER 1.64208613e+02  
163  CharacterType.ARCHER 2.12063259e+02  
150  CharacterType.ARCHER 9.65941194e+01  
89   CharacterType.ARCHER 7.35596012e+01